Canvas with relative width does not repaint correctly on parent resize

cathiechen opened the pull request:
#### 7bbc1c0142c8f349a88396ca519e61599eeec2ec
<pre>
Canvas with relative width does not repaint correctly on parent resize
<a href="https://bugs.webkit.org/show_bug.cgi?id=267986">https://bugs.webkit.org/show_bug.cgi?id=267986</a>

Reviewed by Alan Baradlay.

Elements with relative width do not trigger a full repaint when their size changes because they are not selfNeedsLayout.
For elements like &lt;canvas&gt; and &lt;img&gt;, a size change implies a scale change, so they should trigger a full repaint.
This issue is only revealed if the replaced element is on a tiled layer and its ancestor with selfNeedsLayout on other layer.
If they are on the same layer, the ancestor would trigger full repaint for itself, the repaint rect will cover the replace element.
If the replaced element is on an untiled layer, this layer triggers a full repaint on layer size change.
This patch triggers full repaint for elements like canvas and image in after layout.

* LayoutTests/compositing/tiling/huge-layer-canvas-resize-expected.txt: Added.
* LayoutTests/compositing/tiling/huge-layer-canvas-resize.html: Added.
* LayoutTests/compositing/tiling/huge-layer-img-resize-expected.txt: Added.
* LayoutTests/compositing/tiling/huge-layer-img-resize.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/compositing/tiling/huge-layer-canvas-resize-expected.txt: Added.
* LayoutTests/platform/ios/compositing/tiling/huge-layer-img-resize-expected.txt: Added.
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::issueFullRepaintOnSizeChangeIfNeeded):
(WebCore::RenderReplaced::layout):

Canonical link: <a href="https://commits.webkit.org/294401@main">https://commits.webkit.org/294401@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fae6baa8ed2d9b4ce6b638b270feeaa2c051aca4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101614 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21282 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11596 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106772 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52248 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103654 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21590 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29780 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77386 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34415 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104621 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16672 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91764 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57723 "Found 1 new API test failure: /WPE/TestCookieManager:/webkit/WebKitCookieManager/persistent-storage (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16498 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9781 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51595 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86365 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9858 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109125 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28747 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21160 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86356 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29108 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87965 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85919 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21880 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30669 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8382 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22899 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28675 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33964 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28486 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31809 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30045 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderReplaced.h:
```diff
@@ -51,6 +51,8 @@ class RenderReplaced : public RenderBox {
 
     virtual bool paintsContent() const { return true; }
 
+    virtual bool needsFullRepaintOnSizeChange() const { return false; }
```
Any reason why not just pass in `shouldAlwaysIssueFullRepaint` to `LayoutRepainter` in `RenderReplaced::layout()` when needed?

cathiechen replied:
Yeah, `shouldAlwaysIssueFullRepaint` would work for replaced element without layer. For element hasLayer, `repaintAfterLayoutIfNeeded` is triggered in `recursiveUpdateLayerPositions`, we need to set `NeedsFullRepaint` to its RenderLayer.
I don't have a strong opinion. But if you prefer `shouldAlwaysIssueFullRepaint` + `NeedsFullRepaint` approach, I will adjust the patch later.

alanbaradlay replied:
oh I just realized this requires post layout check. ok, I mean I am fine either way but isn't this just a `repaint()` call which I think could be achieved with a more focused change, something along the lines of

```
static void issueFullRepaintOnSizeChangeIfNeeded(RenderReplaced& renderer)
{
    auto shouldRepaintOnSizeChange = [&] {
        if (is<RenderHTMLCanvas>(renderer))
            return true;
        if (auto* renderImage = dynamicDowncast<RenderImage>(renderer); renderImage && !is<RenderMedia>(*renderImage) && !renderImage->isShowingMissingOrImageError())
            return true;
        return false;
    };
    if (shouldRepaintOnSizeChange())
        renderer.repaint();
}

```
called by `RenderReplaced::layout` when `replacedContentRect() != oldContentRect`

or alternatively `RenderImage/RenderHTMLCanvas::layout` could just call `repaint()` after coming backing from base class layout when content box size changed.

cathiechen replied:
True, it looks clearer.

cathiechen committed:
Canvas with relative width does not repaint correctly on parent resize
https://bugs.webkit.org/show_bug.cgi?id=267986

Reviewed by Alan Baradlay.

Elements with relative width do not trigger a full repaint when their size changes because they are not selfNeedsLayout.
For elements like <canvas> and <img>, a size change implies a scale change, so they should trigger a full repaint.
This issue is only revealed if the replaced element is on a tiled layer and its ancestor with selfNeedsLayout on other layer.
If they are on the same layer, the ancestor would trigger full repaint for itself, the repaint rect will cover the replace element.
If the replaced element is on an untiled layer, this layer triggers a full repaint on layer size change.
This patch triggers full repaint for elements like canvas and image in after layout.

* LayoutTests/compositing/tiling/huge-layer-canvas-resize-expected.txt: Added.
* LayoutTests/compositing/tiling/huge-layer-canvas-resize.html: Added.
* LayoutTests/compositing/tiling/huge-layer-img-resize-expected.txt: Added.
* LayoutTests/compositing/tiling/huge-layer-img-resize.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/compositing/tiling/huge-layer-canvas-resize-expected.txt: Added.
* LayoutTests/platform/ios/compositing/tiling/huge-layer-img-resize-expected.txt: Added.
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::issueFullRepaintOnSizeChangeIfNeeded):
(WebCore::RenderReplaced::layout):

Canonical link: https://commits.webkit.org/294401@main

shallawa commented on LayoutTests/compositing/tiling/huge-layer-canvas-resize.html:
```diff
@@ -0,0 +1,75 @@
+        }
+        if (window.testRunner)
+            testRunner.notifyDone();
+    }, 1000);
```
What exactly are you waiting 1000ms for? This makes the behavior of the test un-deterministic. Besides one second is too long time to wait. Can not it be replaced by 
```
requestAnimationFrame(() => {
    ...
});
```

cathiechen replied:
We need to make sure the repaint rects before css change has flashed, so I added 1000 ms. `requestAnimationFrame` seems not long enough, especially in stress mode.

cathiechen replied:
Rego suggested me to use two `requestAnimationFrame`, it works! I have created a PR to address the code review comments. Please take a look, thanks:) https://github.com/WebKit/WebKit/pull/45040

shallawa commented on LayoutTests/compositing/tiling/huge-layer-img-resize.html:
```diff
@@ -0,0 +1,83 @@
+    img.src = canvas.toDataURL();
+}
+
+async function onImageLoad()
```
Why is this function marked as `async`? It does not have `await` statement.

shallawa commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -132,12 +146,13 @@ void RenderReplaced::layout()
-    if (replacedContentRect() != oldContentRect)
+    if (replacedContentRect() != oldContentRect) {
         setPreferredLogicalWidthsDirty(true);
+        issueFullRepaintOnSizeChangeIfNeeded(*this);
```
I think this is a long name for a function. Could it be just?
```
  if (shouldRepaintOnSizeChange(*this))
        repaint();
```
